homeassistant/components/variables/__init__.py

```diff
@@ -0,0 +1,146 @@
+"""The Variables integration."""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.core import HomeAssistant, ServiceCall, callback
+from homeassistant.helpers import config_validation as cv, entity_registry
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect,
+    async_dispatcher_send,
+)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.storage import Store
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = "variables"
+
+STORAGE_KEY = DOMAIN
+STORAGE_VERSION = 1
+
+DATA_STORAGE = "storage"
+
+TOPIC_VARIABLE_SET = f"{DOMAIN}_set_{0}"
+
+UNIQUE_ID_TEMPLATE = f"{DOMAIN}_{0}"
+
+CONF_KEY = "key"
+CONF_VALUE = "value"
+
+REMOVE_VARIABLE_SCHEMA = vol.Schema({vol.Required(CONF_KEY): cv.string})
+SET_VARIABLE_SCHEMA = vol.Schema(
+    {vol.Required(CONF_KEY): cv.string, vol.Required(CONF_VALUE): cv.string}
+)
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
+
+
+class VariableStore(Store):
+    """Define variable storage."""
+
+    async def _async_migrate_func(self, old_version, old_data):
+        """Migrate to the new version.
+
+        Migrate storage to use format of collection helper.
+        """
+        return old_data
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Variables component."""
+    storage = VariableStore(hass, STORAGE_VERSION, STORAGE_KEY)
+    variables = await storage.async_load() or {}
+
+    entity_component = EntityComponent(_LOGGER, DOMAIN, hass)
+
+    async def async_remove_variable(call: ServiceCall):
+        """Remove a variable."""
+        key = call.data[CONF_KEY]
+
+        if key not in variables:
+            _LOGGER.error("Can't remove non-existent variable: %s", key)
+            return
+
+        _LOGGER.debug("Removing variable: %s", key)
+
+        ent_reg = await entity_registry.async_get_registry(hass)
+        ent_to_remove = ent_reg.async_get_entity_id(
+            DOMAIN, DOMAIN, UNIQUE_ID_TEMPLATE.format(key)
+        )
+        if ent_to_remove is not None:
+            ent_reg.async_remove(ent_to_remove)
+
+        variables.pop(key)
+        hass.async_create_task(storage.async_save(variables))
+
+    async def async_set_variable(call: ServiceCall):
+        """Set a variable."""
+        key = call.data[CONF_KEY]
+        value = call.data[CONF_VALUE]
+
+        if key in variables:
+            _LOGGER.debug('Updating existing variable: %s ("%s")', key, value)
+            async_dispatcher_send(hass, TOPIC_VARIABLE_SET.format(key), value)
+        else:
+            _LOGGER.debug('Creating new variable: %s ("%s")', key, value)
+            await entity_component.async_add_entities([Variable(key, value)])
+
+        variables[key] = value
+        hass.async_create_task(storage.async_save(variables))
```

Comment:
One question is whether it's really necessary to save to storage with _every_ variable update (since those key/value pairs are saved in memory); perhaps we only save on HASS shutdown to avoid a ton of writes?

Reply:
@bachya maybe delay (debounce) writes, so if nothing changes in a minute then write the values. This way will avoid many writes.
Saving only on shutdown can cause the updates to be lost (when the power goes off).
What do you think?